File: src/main/go/leetcode/lists/hot100/LeetCode21.go
package hot100

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {

	dummy := &ListNode{}
	mergeTwoListsHelper(list1, list2, dummy)
	return dummy.Next
}

func mergeTwoListsHelper(list1, list2, node *ListNode) {
	if list1 == nil {
		node.Next = list2
		return
	}
	if list2 == nil {
		node.Next = list1
		return
	}

	if list1 == nil && list2 == nil {
		return
	}
	if list1.Val < list2.Val {
		node.Next = list1
		mergeTwoListsHelper(list1.Next, list2, node.Next)
	} else {
		node.Next = list2
		mergeTwoListsHelper(list1, list2.Next, node.Next)
	}
}

File: src/main/java/leetcode/lists/hot100/LeetCode21.java
package leetcode.lists.hot100;

import common.ListNode;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 23:28
 */
public class LeetCode21 {

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) {
            return list2;
        }
        if (list2 == null) {
            return list1;
        }

        ListNode dummy = new ListNode();
        mergeTwoLists(list1, list2, dummy);
        return dummy.next;

    }

    public ListNode mergeTwoLists(ListNode list1, ListNode list2, ListNode dummy) {
        if (list1 == null && list2 == null) {
            return dummy;
        }
        if (list1 == null) {
            dummy.next = list2;
            return dummy;
        }
        if (list2 == null) {
            dummy.next = list1;
            return dummy;
        }
        if (list1.val <= list2.val) {
            dummy.next = list1;
            return mergeTwoLists(list1.next, list2, dummy.next);
        } else {
            dummy.next = list2;
            return mergeTwoLists(list1, list2.next, dummy.next);
        }
    }
}
